### PR TITLE
Improve performance of massively parallel simulations - updating simulation state / progress bar status 

### LINKED ISSUE #1016: Dropping update_key_hash from from model
Based on https://github.com/yaptide/yaptide/pull/836. update_key_hash in model removed and replaced by JWT token. Change made for doing less db reads/writes.

### PR DESCRIPTION
**Pull request by jagodek**
Based on https://github.com/yaptide/yaptide/pull/836. update_key_hash in model removed and replaced by JWT token. Change made for doing less db reads/writes.

### CODE REVIEW

**Comment by grzanka** [comment]
@jagodek maybe simply skip that Deepsource error:
Function with cyclomatic complexity higher than threshold
PY-R1000

File: yaptide/routes/utils/tokens.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 _Refresh_Token_Expiration_Time = 120  # minutes
 _Access_Token_Expiration_Time = 10  # minutes
 _Keycloak_Token_Expiration_Time = 30  # minutes
+_Simulation_Token_Expiration_time = 7  # days
```
Comment (grzanka) [inline yaptide/routes/utils/tokens.py]: I know there is a comment here mentioning the unit of time. 
All variables above are in minutes, you introduce new one in days.
I'd suggest to be more constistent and follow one of the options
- define all the variables as `timedelta` object: `_Simulation_Token_Expiration_time_days = timedelta(days=7)`
- add suffix to variable which identifies unit of time, like `_Simulation_Token_Expiration_time_days = 7`
- move all the variables to minutes

Personally I am in favour of using timedelta objects here. When you use them its easy to convert them to minutes, seconds of whatever

File: yaptide/routes/utils/tokens.py [hunk 2]
```diff
@@ -37,8 +38,22 @@ def encode_auth_token(user_id: int,
         return e, exp
 
 
-def decode_auth_token(token: str,
-                      is_refresh: bool = False) -> Union[int, str]:
+def encode_simulation_auth_token(simulation_id: int):
+    """Function that encodes JWT token for batched simulation 'update_key'"""
```
Comment (grzanka) [inline yaptide/routes/utils/tokens.py]: by batch we mean submission to Ares HPC, by celery we mean celery queue on local instance. 
This function seems to be used in both scenarios, is that true ?
Adjust the docstring

File: yaptide/routes/utils/tokens.py (hunk 1 above)
Comment (jagodek) [reply yaptide/routes/utils/tokens.py]: It's a change made by original author of these changes so i didn't really noticed but you right about those days.

File: yaptide/routes/utils/tokens.py (hunk 2 above)
Comment (jagodek) [reply yaptide/routes/utils/tokens.py]: Don't know what was the intention here. I'll edit docstring as sugested

**Comment by grzanka** [comment]
@jagodek please rebase/update and describe in the issue what is the real problem or what kind of improvement we propose

testing is needed

**Comment by grzanka** [comment]
@jagodek can you fix the conflicts, probably related to merging of https://github.com/yaptide/yaptide/pull/1077 ?

**Comment by grzanka** [comment]
@jagodek conflicts again...

File: yaptide/persistence/models.py [hunk 3]
```diff
@@ -92,23 +92,14 @@ class SimulationModel(db.Model):
                                        nullable=False,
                                        default=EntityState.UNKNOWN.value,
                                        doc="Simulation state (i.e. 'pending', 'running', 'completed', 'failed')")
-    update_key_hash: Column[str] = db.Column(db.String,
```
Comment (grzanka) [inline yaptide/persistence/models.py:98]: now as we have flask migrate this can be done safer !

**Comment by jagodek** [comment]
I've run 4 tests for each branch in configuration: 
Example: Proton beam scattering on lead slab
particles: 1e+7
number of tasks:  100 
average time for this branch was  4:55 while for master 4:59

**Comment by grzanka** [comment]
> I've run 4 tests for each branch in configuration: Example: Proton beam scattering on lead slab particles: 1e+7 number of tasks: 100 average time for this branch was 4:55 while for master 4:59

Have you checked CPU usage of flask container for both cases ?

**Comment by jagodek** [comment]
> > I've run 4 tests for each branch in configuration: Example: Proton beam scattering on lead slab particles: 1e+7 number of tasks: 100 average time for this branch was 4:55 while for master 4:59
> 
> Have you checked CPU usage of flask container for both cases ?

oh. no. I'll try again then.

**Comment by jagodek** [comment]
image
so this is the improve on cpu

**Comment by grzanka** [comment]
@jagodek thats exactly the improvement. Can you fix the tests and handle flask-migrate stuff ?

**Comment by grzanka** [comment]
We can safely delete `update_column_hash` from the prod DB - its only needed for running simulations.

**Comment by jagodek** [comment]
done

**Comment by grzanka** [comment]
> done

may I do some tests on yap-dev ?

**Comment by jagodek** [comment]
> > done
> 
> may I do some tests on yap-dev ?

yes

**Comment by grzanka** [comment]
@jagodek have you checked if cancelling simulations works when working with Celery ?

I've submitted longer simulation on 10 tasks (Celery) and got follwing error from Flask when trying to cancel it.
BTW - I deleted `yaptide_data` volume prior to this test to make a check on fresh DB.

```
172.18.0.6 - - [29/Oct/2024 17:08:07] "OPTIONS /jobs/direct?job_id=20241029-170747-4d2942af-df58-4593-bac5-3a5584c107bfDIRECT HTTP/1.0" 200 -
INFO:werkzeug:172.18.0.6 - - [29/Oct/2024 17:08:07] "OPTIONS /jobs/direct?job_id=20241029-170747-4d2942af-df58-4593-bac5-3a5584c107bfDIRECT HTTP/1.0" 200 -
[2024-10-29 17:08:08,019] ERROR in app: Exception on /jobs/direct [DELETE]
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_restful/__init__.py", line 489, in wrapper
    resp = resource(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/views.py", line 110, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_restful/__init__.py", line 604, in dispatch_request
    resp = meth(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/app/yaptide/routes/utils/decorators.py", line 24, in wrapper
    return f(user, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/app/yaptide/routes/celery_routes.py", line 162, in delete
    result: dict = cancel_job(merge_id=simulation.merge_id, celery_ids=celery_ids)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/app/yaptide/celery/utils/manage_tasks.py", line 83, in cancel_job
    "tasks": [cancel_task(job_id, "task_state") for job_id in celery_ids]
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/app/yaptide/celery/utils/manage_tasks.py", line 65, in cancel_task
    job_state: str = translate_celery_state_naming(job.state)
                                                   ^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/celery/result.py", line 503, in state
    return self._get_task_meta()['status']
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/celery/result.py", line 442, in _get_task_meta
    return self._maybe_set_cache(self.backend.get_task_meta(self.id))
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/celery/backends/base.py", line 608, in get_task_meta
    meta = self._get_task_meta_for(task_id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/celery/backends/base.py", line 995, in _get_task_meta_for
    meta = self.get(self.get_key_for_task(task_id))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/celery/backends/base.py", line 869, in get_key_for_task
    raise ValueError(f'task_id must not be empty. Got {task_id} instead.')
ValueError: task_id must not be empty. Got  instead.
ERROR:yaptide:Exception on /jobs/direct [DELETE]
```

**Comment by jagodek** [comment]
I see there's problem. But i notice same thing on master.

**Comment by grzanka** [comment]
> I see there's problem. But i notice same thing on master.

Interesting, I wasn't able to spot this problem on master (after cleaning the DB with `docker volume rm yaptide_data` once I finished the test on this PR).

**Comment by grzanka** [comment]
This is a test on master (but with a fresh DB):
[Screencast from 2024-10-29 19-42-49.webm](https://github.com/user-attachments/assets/9cb0205c-3460-4bc3-a174-712457f3979a)

**Review by grzanka — APPROVED** [review]

**Pull request merged**